uprotocol/client/usubscription/v3/inmemoryusubcriptionclient.py

```diff
@@ -218,11 +245,119 @@ async def unregister_listener(self, topic: UUri, listener: UListener) -> UStatus
             raise ValueError("Unsubscribe topic missing")
         if not listener:
             raise ValueError("Request listener missing")
-        return await self.transport.unregister_listener(topic, listener)
+        status = await self.transport.unregister_listener(topic, listener)
+        self.handlers.pop(UriSerializer.serialize(topic), None)
+        return status
 
     def close(self):
         """
         Close the InMemoryRpcClient by clearing stored requests and unregistering the listener.
         """
         self.handlers.clear()
         self.notifier.unregister_notification_listener(self.notification_uri, self.notification_handler)
+
+    async def register_for_notifications(
+        self, topic: UUri, handler: SubscriptionChangeHandler, options: Optional[CallOptions] = CallOptions.DEFAULT
+    ):
+        """
+        Register for Subscription Change Notifications.
+
+        This API allows producers to register to receive subscription change notifications for
+        topics that they produce only.
+
+        :param topic: UUri, The topic to register for notifications.
+        :param handler: callable, The SubscriptionChangeHandler to handle the subscription changes.
+        :param options: CallOptions, The CallOptions to be used for the register request.
+
+        :return: asyncio.Future[NotificationsResponse], A future object that completes with NotificationsResponse
+            if the uSubscription service accepts the request to register the caller to be notified of subscription
+            changes, or raises an exception if there is a failure reason.
+        """
+        if not topic:
+            raise ValueError("Topic missing")
+        if not handler:
+            raise ValueError("Handler missing")
+        if not options:
+            raise ValueError("CallOptions missing")
+
+        request = NotificationsRequest(topic=topic, subscriber=SubscriberInfo(uri=self.transport.get_source()))
+
+        response = self.rpc_client.invoke_method(self.register_for_notification_uri, UPayload.pack(request), options)
+        notifications_response = await RpcMapper.map_response(response, NotificationsResponse)
+        if handler:
+            topic_str = UriSerializer.serialize(topic)
+            if topic_str in self.handlers and self.handlers[topic_str] != handler:
+                raise UStatusError.from_code_message(UCode.ALREADY_EXISTS, "Handler already registered")
+            self.handlers[topic_str] = handler
```

Comment:
Curious why you serialize to string for your maps, is string comparison faster than object comparison? Also what if the topic has a wildcard in it, I guess it wouldn't work?

Reply:
In Python, dictionary keys must be hashable. Since UUri objects are not hashable by default, using them directly as dictionary keys will raise an `unhashable type: UUri` exception. To address this, we can either serialize the UUri objects and store the serialized data, or create a wrapper class that implements the `__hash__` and `__eq__` methods. I found the first option easier.